continued support for noir -> r1cs

dolores-park opened the pull request:
- add r1cs optimization, 1 poseidon row count reduced from 644->428

recmo commented on noir-r1cs/src/compiler.rs:
```diff
@@ -116,19 +119,31 @@ impl R1CS {
+            let term = expr.mul_terms[0];
+            a = vec![(FieldElement::one(), self.map_witness(term.1))];
+            b = vec![(FieldElement::one(), self.map_witness(term.2))];
+        }
```
There is a more general result: The last mul term never needs an intermediate constraint (even if there are multiple mul-terms).

dolores-park commented:
@recmo 
having a little trouble confirming that the witness generation process is correct.
try to run
`cargo run -- noir /Users/ende.shen/pork/ProveKit/noir-r1cs/noir-examples/basic-2/target/basic.json /Users/ende.shen/pork/ProveKit/noir-r1cs/noir-examples/basic-2/target/basic.gz` 
or 
`cargo run -- noir /Users/ende.shen/pork/ProveKit/noir-r1cs/noir-examples/basic/target/basic.json /Users/ende.shen/pork/ProveKit/noir-r1cs/noir-examples/basic/target/basic.gz`
after running nargo execute on basic and basic-2
think the solving process isn't completed
I will look into this, and also just flagging it so you know

dolores-park commented:
@recmo think it's a good place to merge into main, would love to get feedback / approval from you
for the updates:
- fixed & optimized the -> r1cs process
- added witness population from .gz witness files
- added bunch of IO functions
to test,
run commands like
`cargo run -- noir ProveKit/noir-r1cs/noir-examples/poseidon-rounds/target/basic.json ProveKit/noir-r1cs/noir-examples/poseidon-rounds/target/basic.gz`
after that,
there should be a `r1cs.json` file generated,
and that file can be copied into `ssh://git@github.com/reilabs/ProveKit.git`, and proven with their `prover/src/main.rs`

dolores-park commented:
Added some scripts from @RyanCao7 , to run use
```
cd noir-r1cs
bash scripts/run_all.sh
```

dolores-park commented:
To have a full end to end run,
try picking a noir program
say `cd noir-r1cs/noir-examples/poseidon-rounds`
provide the correct `Prover.toml` file under that dir (inputs to noir programs)
and run
`nargo execute`
to generate
`.json` file -- represent the circuit represented in ACIR
and `.gz` file -- represent the witness

after having the noir intermediate representations, compile file at `noir-r1cs/src/main.rs`
and run
`target/debug/noir-r1cs /Users/ende/modulus/ProveKit/noir-r1cs/noir-examples/poseidon-rounds/target/basic.json /Users/ende/modulus/ProveKit/noir-r1cs/noir-examples/poseidon-rounds/target/basic.gz`
the two args are path to `.json` and path to `.gz`

after that, there should be a `r1cs.json` at `./ProveKit`, this is the ABC matrices that representing the r1cs

==========================================================================================
        (For these below, also checkout https://github.com/reilabs/gnark-whir/blob/main/README.md)

copy that to the repo at
`https://github.com/reilabs/ProveKit.git`
as `prover/r1cs.json`
and run
`rustup run beta cargo run --bin prover -- --input_file_path prover/r1cs.json`
to get the two files
`prover/params` and `prover/proof`

checkout the repo at
`https://github.com/reilabs/gnark-whir`
cd into the repo, and run
`go run .`
note that it assumes 
`https://github.com/reilabs/gnark-whir` and `https://github.com/reilabs/ProveKit.git` live side by side, and the two files `prover/params` and `prover/proof` live in `https://github.com/reilabs/ProveKit.git`.

==========================================================================================

recmo commented:
For above instructions to work it is important that the `noirc` version matches the `noirc_artifacts` crate. 

```sh
noirup -C 03b58fa2
```
